articles/hajimehoshi.re

```diff
@@ -0,0 +1,102 @@
+= @<tt>{gomobile}完全に理解した
+
+完全に理解したとは言っていない。
+
+== 著者について
+
+星一 (@<tt>{@hajimehoshi}) と申します。
+趣味でGoを使っています。
+代表作はEbiten@<fn>{ebiten}という2Dゲームライブラリです。
+Ebitenと@<tt>{gomobile}を使ったゲームをいくつか共同制作していて、2019年8月現在の代表作は「くまのレストラン@<fn>{bearsrestaurant}」です。
+
+//footnote[ebiten][@<href>{https://ebiten.org}]
+//footnote[bearsrestaurant][@<href>{https://daigostudio.com/bearsrestaurant}]
+
+== @<tt>{gomobile}
+
+@<tt>{gomobile}はAndroidおよびiOS向けのアプリケーションもしくはライブラリを出力するソフトウェアです。
+パッケージは@<tt>{golang.org/x/mobile}です。
+
+なぜ@<tt>{gomobile}がGo本体以外に必要になるかと言うと、Go本体ではモバイル向けのアプリケーションやライブラリを作れないからです。
+Go本体では、AndroidとiOS向けに、C互換の共有ライブラリまたは静的ライブラリを出力することができます。
+実際@<tt>{gomobile}はGo本体にコンパイル処理を丸投げしています。
+@<tt>{gomobile}はGoのコードを単にコンパイルするだけではなく、AndroidやiOSアプリケーションとして使いやすい形に変換します。
+たとえばAndroidが使えるライブラリは@<tt>{.aar}形式であり、特定のファイルをzipで固めたものになりますが、
+Go本体はこれを生成してくれません。
+また、たとえばGoの関数をJavaで使う際には、バインディングのためのコードが別途必要になってしまいます。
+@<tt>{gomobile}はそれらをやっているわけです。
+
+@<tt>{gomobile}コマンドは大雑把に2つのモードがあります。
+@<tt>{gomobile build}と@<tt>{gomobile bind}です。
+
+@<tt>{gomobile build}は単独のアプリケーションを出力します。
+Android向けには@<tt>{.apk}ファイル、iOS向けには@<tt>{.ipa}ファイルです。
+アプリケーション全体をすべてGoで書くことが可能です。
+機能は限定されていて、アプリケーションはOpenGLサーフェスを一個だけ持ちます。
+
+一方@<tt>{gomobile bind}はライブラリを出力します。
+Android向けには@<tt>{.aar}ファイル、iOS向けには@<tt>{.framework}ファイルです。
+Goのexportされた関数などは、一定の条件@<fn>{gobind}を満たしたものは、JavaもしくはObjective-C側でも呼び出せるようになります。
+Goで定義したinterfaceをJava側で実装してGoの関数に渡す、といったことも可能です。
+あくまで共有ライブラリなので、これ単独では完成したアプリケーションになりません。
+ユーザーはGo以外の言語を書く必要があります。
+
+//footnote[gobind][https://godoc.org/golang.org/x/mobile/cmd/gobind]
+
+@<tt>{gomobile build}は簡単に完成されたアプリケーションを作れる一方、柔軟性に欠けます。
+@<tt>{gomobile build}で作られたアプリケーションはOpenGLサーフェス一個だけしかもっておらず、
+他のコンポーネントを入れることができません。またストアに必要な署名などのファイルを挿入するのは、
+不可能ではないにしろ困難です。
+@<tt>{gomobile build}だけで、ストアに載せる実用アプリケーションを作ることは現実的ではありません@<fn>{tenntenn}。
+実際、我々が開発しているゲームでは@<tt>{gomobile bind}を活用しています。
+
+//footnote[tenntenn][実際に@<tt>{gomobile build}で作られたアプリケーションをストアに載せた例もあります: @<href>{https://www.slideshare.net/takuyaueda967/go-mobileandroid}]
```

Comment:
**[textlint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Line 53 sentence length(103) exceeds the maximum sentence length of 100.
Over 3 characters. (ja-technical-writing/sentence-length)